File: python/0219_contains_duplicate_II.py
# n = len(nums)
# Time: O(n)
# Space: O(n)
from typing import List


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        num_positions = {}

        for i, val in enumerate(nums):
            if val in num_positions and i - num_positions[val] <= k:
                return True
            else:
                num_positions[val] = i

        return False

File: go/0219_contains_duplicate_II.go
package main

// n = len(nums)
// Time: O(n)
// Space: O(n)
func containsNearbyDuplicate(nums []int, k int) bool {
	numsIndex := make(map[int]int)

	for cur_idx, num := range nums {
		if prev_idx, exist := numsIndex[num]; exist {
			if cur_idx-prev_idx <= k {
				return true
			}
		}
		numsIndex[num] = cur_idx
	}

	return false
}
